### ISSUE TITLE
java.lang.IndexOutOfBoundsException: toIndex = 3

### ISSUE THREAD

**Issue by JayJiangCT** [mapbox/mapbox-navigation-android#5157]
**Android API:** **30**
**Mapbox Navigation SDK version:** **2.0.2**

### Steps to trigger behavior

 1. Plan a navigation route and start the navigation（orginal waypoint is [40.661087,-74.354304], destination waypoint is [40.658842,-74.35648], route.geometry is "geometry=yuvplArmeylCl^p_@gRfa@jhAjjAr_@yu@za@~b@jOnLrEhAlh@lArDj@jDjBvh@ff@cw@ryAc`BgdB"）
 2. Drive on the alden street in westfiled NJ.Drive along the planned route at first, when you reach the first intersection, turn right and go straight along Highland. A rerouting will happen at this time. Soon I got a new navigation route, then I continue driving along the given route. When I don’t follow the given route, a new rerouting occurs. Then the app crashes.
 3. I checked the logs and found stack information like this：
```
java.lang.IndexOutOfBoundsException: toIndex = 3
        at java.util.ArrayList.subListRangeCheck(ArrayList.java:1016)
        at java.util.ArrayList.subList(ArrayList.java:1008)
        at com.mapbox.navigation.core.routeoptions.RouteOptionsUpdater.update(RouteOptionsUpdater.kt:102)
        at com.mapbox.navigation.core.reroute.MapboxRerouteController.reroute(MapboxRerouteController.kt:64)
        at com.mapbox.navigation.core.MapboxNavigation.reroute(MapboxNavigation.kt:1222)
        at com.mapbox.navigation.core.MapboxNavigation.createInternalOffRouteObserver$lambda-5(MapboxNavigation.kt:1156)
        at com.mapbox.navigation.core.MapboxNavigation.$r8$lambda$Fl3MnjuCU8BFshEUTwlLmlSZqEU(Unknown Source:0)
        at com.mapbox.navigation.core.MapboxNavigation$$ExternalSyntheticLambda2.onOffRouteStateChanged(Unknown Source:2)
        at com.mapbox.navigation.core.trip.session.MapboxTripSession.setOffRoute(MapboxTripSession.kt:149)
        at com.mapbox.navigation.core.trip.session.MapboxTripSession.access$setOffRoute(MapboxTripSession.kt:53)
        at com.mapbox.navigation.core.trip.session.MapboxTripSession$navigatorObserver$1$onStatus$1.invokeSuspend(MapboxTripSession.kt:297)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:246)
```

### Expected behavior
Plan a new route and continue to navigate

### Actual behavior
App crashes

### Problem inference
```
try {
            routeProgress.currentLegProgress?.legIndex?.let { index ->
                val location = locationMatcherResult.enhancedLocation
                optionsBuilder
                    .coordinatesList(
                        coordinatesList
                            .drop(coordinatesList.size - remainingWaypoints).toMutableList().apply { //this remainingWaypoints's value 
is wrong, it will case the size of coordinatesList will be larger than original.
                                add(0, Point.fromLngLat(location.longitude, location.latitude))
                            }
                    )
                    .bearingsList(
                        getUpdatedBearingList(
                            coordinatesList.size,
                            location.bearing.toDouble(),
                            routeOptions.bearingsList(),
                            remainingWaypoints
                        )
                    )
                    .radiusesList(
                        let radiusesList@{
                            val radiusesList = routeOptions.radiusesList()
                            if (radiusesList.isNullOrEmpty()) {
                                return@radiusesList emptyList<Double>()
                            }
                            mutableListOf<Double>().also {
                                it.addAll(radiusesList.subList(index, coordinatesList.size))
                            }
                        }
                    )
                    .approachesList(
                        let approachesList@{
```
                            val approachesList = routeOptions.approachesList()
                            if (approachesList.isNullOrEmpty()) {
                                return@approachesList emptyList<String>()
                            }
                            mutableListOf<String>().also {
                                it.addAll(approachesList.subList(index, coordinatesList.size))
                            }
                        }
                    )
The value returned by **remainingWaypoints** is wrong. 

https://user-images.githubusercontent.com/59813406/143164369-d0831b7f-0dc2-4708-b63d-6173c5ea6da5.mp4

**Comment by VysotskiVadim** [comment]
@JayJiangCT, is the bug reproducible in your application? Can you reproduce it again? If yes, can you please share the logs. Logs should contain requests to Directions API which could help troubleshooting. Please also try to record a history file using `MapboxNavigation.historyRecorder`.

I tried to reproduce on the example from the SDK repo using your instructions, but everything works fine for me. Please check the video and verify if I followed all steps you mentioned.

https://user-images.githubusercontent.com/6190346/144209845-e7e980a6-3352-422e-8d64-862c805170e5.mp4

**Comment by JayJiangCT** [comment]
@VysotskiVadim, thank you very much for your reply. I think the reason why your application did not crash is that you did not add the relevant parameters of `approach `in the requested route. I'll show you all parameters in the direction api.

**Comment by JayJiangCT** [comment]
In order to circumvent this problem, we have replaced the default RerouteController in your SDK, but I still hope that you can fix this problem.

**Comment by JayJiangCT** [comment]
Awesome~

**Comment by VysotskiVadim** [comment]
The fix is available in the version [v2.2.0-beta.1](https://github.com/mapbox/mapbox-navigation-android/releases/tag/v2.2.0-beta.1).
